### PR TITLE
Used addCleanup() in tests where appropriate.

### PR DESCRIPTION
**Pull request by felixxm**

### CODE REVIEW

**Comment by felixxm** [comment]
This is the first iterations. Where appropriate, I'd like to switch to `enterClassContext` in the follow-up PR.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Review by hramezani — APPROVED** [review]

File: tests/file_storage/tests.py [hunk 1]
```diff
@@ -1018,11 +1012,9 @@ class FileStoragePermissions(unittest.TestCase):
     def setUp(self):
         self.umask = 0o027
         self.old_umask = os.umask(self.umask)
+        self.addCleanup(os.umask, self.old_umask)
```
Comment (claudep) [inline tests/file_storage/tests.py:1015]: Nit-pick: old_umask doesn't need to be defined on self. There are some other occurrences in this file where the defined variable is not used later in tests, so does not need to be on self. Maybe you prefer consistency and always define the variable on self?

File: tests/transactions/tests.py [hunk 2]
```diff
@@ -249,22 +249,18 @@ class AtomicInsideTransactionTests(AtomicTests):
     def setUp(self):
         self.atomic = transaction.atomic()
         self.atomic.__enter__()
-
-    def tearDown(self):
-        self.atomic.__exit__(*sys.exc_info())
+        self.addCleanup(self.atomic.__exit__, *sys.exc_info())
```
Comment (claudep) [inline tests/transactions/tests.py:252]: I suspect `sys.exc_info()` is not necessarily identical in `setUp` and in `tearDown`. Don't know if this matters or not.

**Review by claudep — APPROVED** [review]
Mostly OK, just some comments, that may be ignored.

File: tests/file_storage/tests.py [hunk 3]
```diff
@@ -71,16 +71,14 @@ class FileStorageTests(SimpleTestCase):
...
             location=self.temp_dir, base_url="/test_media_url/"
         )
         # Set up a second temporary directory which is ensured to have a mixed
         # case name.
         self.temp_dir2 = tempfile.mkdtemp(suffix="aBc")
-
-    def tearDown(self):
-        shutil.rmtree(self.temp_dir)
-        shutil.rmtree(self.temp_dir2)
+        self.addCleanup(shutil.rmtree, self.temp_dir2)
```
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/file_storage/tests.py:81]: Good catch :+1:

File: tests/transactions/tests.py (hunk 2 above)
Comment (felixxm) [reply tests/transactions/tests.py:252]: Let's revert :+1:

File: tests/file_storage/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/file_storage/tests.py]: Updated :+1: I've tried to remove `self` when possible.

**Comment by felixxm** [comment]
Thanks both for reviews :+1:

**Review by ngnpope — APPROVED** [review]
Had a quick look through and this looks fine. I have some observations though.

I see you're looking to use `enterClassContext()` which is great, but that is for use in `setUpClass()`. This PR is focusing on `setUp()` for which you'll want to backport `enterContext()` which I didn't do when we introduced `enterClassContext()` in ed4f83782d9f3404ad600f6131ef78244ff1e162 [^1]:

```patch
diff --git a/django/test/testcases.py b/django/test/testcases.py
index 0e887bdc0b..4530779449 100644
--- a/django/test/testcases.py
+++ b/django/test/testcases.py
@@ -337,6 +337,10 @@ def _post_teardown(self):
         def enterClassContext(cls, cm):
             return _enter_context(cm, cls.addClassCleanup)
 
+        # Backport of unittest.TestCase.enterContext() from Python 3.11.
+        def enterContext(self, cm):
+            return _enter_context(cm, self.addCleanup)
+
     def settings(self, **kwargs):
         """
         A context manager that temporarily sets a setting and reverts to the
```

I think it's also worth following up with:

- Use `addClassCleanup()` in tests where appropriate.
- Convert from `setUp()` to `setUpClass()` where per-test creation/deletion is unnecessary.
- Make use of `enterContext()` or `enterClassContext()` where viable.

[^1]: Also note that there is an `enterModuleContext()`, but I suggest we don't bother with that as it'd mean having to import that from `django.test` and Django doesn't seem to have any use of `setUpModule()` or `tearDownModule()` anyway.

**Comment by felixxm** [comment]
@ngnpope Thanks :+1: 
> I see you're looking to use `enterClassContext()` which is great, but that is for use in `setUpClass()`.

Yes I know, but, as far as I'm aware, most of `enable`/`disable` round-trips for patched settings could be moved to the `setUpClass()` and `enterClassContext()`. At least that was my initial idea.

**Comment by felixxm** [comment]
I'll prepare follow up PRs in the New Year :fireworks:

**Comment by felixxm** [comment]
Small follow up #17673.

**Pull request merged**